consensus/core/src/synchronizer.rs

```diff
@@ -749,50 +737,80 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
-
-                                            total_stake += context.committee.stake(authority_index);
-                                        },
-                                        Err(err) => {
-                                            warn!("Invalid result returned from {authority_index} while fetching last own block: {err}");
+                // Get the highest of all the results. Retry until at least `f+1` results have been gathered.
+                let mut total_stake;
+                let mut highest_round;
+                let mut retries = 0;
+                let mut retry_delay_step = Duration::from_millis(500);
+                'main:loop {
+                    total_stake = 0;
+                    highest_round = 0;
+
+                    // Ask all the other peers about our last block
+                    let mut results = FuturesUnordered::new();
+
+                    for (authority_index, _authority) in context.committee.authorities() {
+                        if authority_index != context.own_index {
+                            results.push(fetch_own_block(authority_index, Duration::from_millis(0)));
+                        }
+                    }
+
+                    // Gather the results but wait to timeout as well
+                    let timer = sleep_until(Instant::now() + context.parameters.sync_last_known_own_block_timeout);
+                    tokio::pin!(timer);
+
+                    'inner: loop {
+                        tokio::select! {
+                            result = results.next() => {
+                                let Some((result, authority_index)) = result else {
+                                    break 'inner;
+                                };
+                                match result {
+                                    Ok(result) => {
+                                        match process_blocks(result, authority_index) {
+                                            Ok(blocks) => {
+                                                let max_round = blocks.into_iter().map(|b|b.round()).max().unwrap_or(0);
+                                                highest_round = highest_round.max(max_round);
+
+                                                total_stake += context.committee.stake(authority_index);
+                                            },
+                                            Err(err) => {
+                                                warn!("Invalid result returned from {authority_index} while fetching last own block: {err}");
+                                            }
                                         }
+                                    },
+                                    Err(err) => {
+                                        warn!("Error {err} while fetching our own block from peer {authority_index}. Will retry.");
+                                        results.push(fetch_own_block(authority_index, FETCH_OWN_BLOCK_RETRY_DELAY));
```

Comment:
Should we add a delay before the retry?

Reply:
The `FETCH_OWN_BLOCK_RETRY_DELAY` is the delay.  `fetch_own_block` inside will sleep for `FETCH_OWN_BLOCK_RETRY_DELAY` before try to fetch again the block. Do you mean something different?